### PR TITLE
feat(notification): Clickable notifications

### LINKED ISSUE #694: task(notifications): Clicking banner alert should navigate user to said chat
**Task:**
When you are clicked out of focus of the app and you receive a new message/get the banner alert in the top right of the screen, you should be able to click that alert and have it take you to said chat 

**OS:**
All

**Screenshot:**

Screenshot 2023-04-20 at 3 18 03 PM

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Makes it so notifications can be used to run actions inside the app:
  -  When you tap out of the app receiving new messages or friend requests makes a notification pop up
  - If that notification is a new message clicking on it will bring you to the chat where the message came from
  - If that notification is a friend request clicking on it will bring you to the friend request page
- On linux the actions are a dropdown
Screenshot from 2023-05-05 14-05-10
- ~~Needs testing on macos~~
- ~~Windows is missing atm~~
- Windows:
  - Windows Toast need to have an AppID assigned for the toasts. Currently it uses Powershell here so the toast shows that it comes from powershell. To be a valid AppID it needs to be also a shortcut in the startmenu. Im not sure what our plan regarding that will be but a possible way of doing is outlined here:
  [https://smsagent.blog/2020/10/20/adding-your-own-caller-app-for-custom-windows-10-toast-notifications/](https://smsagent.blog/2020/10/20/adding-your-own-caller-app-for-custom-windows-10-toast-notifications/)
  Changing the app id would then be done here:
  [https://github.com/Satellite-im/Uplink/pull/786/files#diff-3a3b0114c99d74ea1fc510bd6ed952782740c102efea7fbe4471830949776b7dR129](https://github.com/Satellite-im/Uplink/pull/786/files#diff-3a3b0114c99d74ea1fc510bd6ed952782740c102efea7fbe4471830949776b7dR129)

### Which issue(s) this PR fixes 🔨

- Resolve #694

### CODE REVIEW

**Comment by phillsatellite** [comment]
Hi @Flemmli97 👋  I tested on MacOS (and linux and Windows even though I know you said Windows wasnt working yet) but I haven't been able to get them to work on any OS. I tried grabbing a screen recording but when you are recording notifications are silenced. 

but on MacOS I tried clicking notification and I also tried clicking _show_ to see if that would work and no luck on either 

Screenshot 2023-05-05 at 10 46 02 AM

**Comment by Flemmli97** [comment]
> Hi @Flemmli97 wave I tested on MacOS (and linux and Windows even though I know you said Windows wasnt working yet) but I haven't been able to get them to work on any OS. I tried grabbing a screen recording but when you are recording notifications are silenced.
> 
> but on MacOS I tried clicking notification and I also tried clicking _show_ to see if that would work and no luck on either
> 
> Screenshot 2023-05-05 at 10 46 02 AM

how did it go on linux?

**Review by stavares843 — COMMENTED** [review]
doesn't build here

image

macOS, m1

**Comment by phillsatellite** [comment]
I know its just a draft but wanted to test it out again

Linux and Mac I was able to get to work 

https://user-images.githubusercontent.com/93608357/236854059-21e9987e-9e70-4602-b008-1688a6768c4d.mov

But on WindowsVM I was not able to compile 

Screenshot 2023-05-08 at 10 40 38 AM

**Comment by sdwoodbury** [comment]
https://github.com/Satellite-im/Uplink/assets/8636602/6225cd9e-bb1a-41d2-aed0-873dc9443626

I can compile but it appears that clicking the notification doesn't switch windows. then again, maybe this is related to me having 2 instances of Uplink running on the same VM.

**Comment by stavares843** [comment]
linter is failing 🔨

**Comment by phillsatellite** [comment]
When tested on MacOS: User will receive notification, when they click alert it will not take the user into the app. Within the app you can see it does go to the correct chat page but it does click the user into the app. 

https://github.com/Satellite-im/Uplink/assets/93608357/0439e985-494b-4485-aa79-59cd152660f8


WindowsOS: The notification will not appear for me on Windows. Sound comes in fine but no alert appears

**Comment by Flemmli97** [comment]
> When tested on MacOS: User will receive notification, when they click alert it will not take the user into the app. Within the app you can see it does go to the correct chat page but it does click the user into the app.
> 
>  Screen.Recording.2023-05-22.at.12.36.15.PM.mov 
> WindowsOS: The notification will not appear for me on Windows. Sound comes in fine but no alert appears

if you use the notification on dev branch on mac does it open the app?
and strange that windows doesnt work. it did for me

**Comment by phillsatellite** [comment]
> > When tested on MacOS: User will receive notification, when they click alert it will not take the user into the app. Within the app you can see it does go to the correct chat page but it does click the user into the app.
> > Screen.Recording.2023-05-22.at.12.36.15.PM.mov
> > WindowsOS: The notification will not appear for me on Windows. Sound comes in fine but no alert appears
> 
> if you use the notification on dev branch on mac does it open the app? and strange that windows doesnt work. it did for me

On dev branch clicking the notification does not open up the app, but it should not since we haven't merged this in yet

**Comment by Flemmli97** [comment]
> > > When tested on MacOS: User will receive notification, when they click alert it will not take the user into the app. Within the app you can see it does go to the correct chat page but it does click the user into the app.
> > > Screen.Recording.2023-05-22.at.12.36.15.PM.mov
> > > WindowsOS: The notification will not appear for me on Windows. Sound comes in fine but no alert appears
> > 
> > 
> > if you use the notification on dev branch on mac does it open the app? and strange that windows doesnt work. it did for me
> 
> On dev branch clicking the notification does not open up the app, but it should not since we haven't merged this in yet

ok. i just wanted to make sure whether it auto does it already and this pr breaks that or not as on linux it opens the app by default

**Review by phillsatellite — APPROVED** [review]

File: ui/src/layouts/friends.rs
Comment (sdwoodbury) [inline ui/src/layouts/friends.rs]: non-blocking: I believe that `initial_page` doesn't need to be a reference and that you can just clone the `UseRef` when passing it to `Props`. Doing so would remove the need for a generic both in `friends::Props` and in `FriendsLayout`.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: seems that the wrong message is being passed to `error!`.  The command isn't used to initialize warp. The channel goes to `main.rs > notification_action_handler` and all it does is replace routes.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: seems that the wrong message is being passed to `error!`.  The command isn't used to initialize warp. The channel goes to `main.rs > notification_action_handler` and all it does is replace routes.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: it's probably better to not `unwrap` but instead return an error.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: it's probably better to not `unwrap` but instead return an error.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: it's probably better to not `unwrap` but instead return an error.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: it's probably better to not `unwrap` but instead return an error.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: instead of passing a String to the toast_notification, would it be sufficient to simply attach a callback (or in the case of MacOs, process the NotificationResponse) that sends the `NotificationAction` over the channel? It seems like that would be enough. That would remove the need for the `action_id` argument in the `show_with_action` function. That variable seems redundant anyway - just passing the `NotificationAction` should be enough. If something is needed for logging, `derive_more::Display` can be derived by `NotificationAction`.

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: the message being passed to `trace!` seems inaccurate. the thing being "activated" is a toast notification, not a "windows activation"

File: common/src/notifications.rs
Comment (sdwoodbury) [inline common/src/notifications.rs]: the name `push_notification_actionable` may be better as `push_desktop_notification`. That is the difference - desktop notifications are being created here, while `push_notification` results in a toast being displayed only within Uplink.

**Review by sdwoodbury — CHANGES_REQUESTED** [review]
Surprise - I decided to go through the pull requests on Uplink today :grin: 

in addition to the below comments
- It seems that `push_notification` is no longer used. Before I suggest we remove it i wanted to ask - if the app is focused, is it better to launch a desktop notification or a toast within the app? if the latter is preferred, then maybe we can add a `None` enum to `NotificationAction` and have `push_notification` decide which type of notification to send (toast or desktop notification). 
- I found clippy warnings when I ran `clippy` on a mac. The cargo workflow didn't see them because they are only run for Linux.

File: common/src/notifications.rs
Comment (Flemmli97) [reply common/src/notifications.rs]: the notification for uplink only isnt generated via `push_notification`. that is handled somewhere else. `push_notification` was used for desktop notification but without action handling

adding on to your initial comment: i left it there in case there might be a need for it without actions but if not that can be removed

File: common/src/notifications.rs
Comment (Flemmli97) [reply common/src/notifications.rs]: resolved in [fafcaed](https://github.com/Satellite-im/Uplink/pull/786/commits/fafcaed59113632974693c4afed9af931faa0c69)

File: common/src/notifications.rs
Comment (Flemmli97) [reply common/src/notifications.rs]: resolved in [fafcaed](https://github.com/Satellite-im/Uplink/pull/786/commits/fafcaed59113632974693c4afed9af931faa0c69)

File: common/src/notifications.rs
Comment (Flemmli97) [reply common/src/notifications.rs]: resolved in [fafcaed](https://github.com/Satellite-im/Uplink/pull/786/commits/fafcaed59113632974693c4afed9af931faa0c69)

File: ui/src/layouts/friends.rs
Comment (Flemmli97) [reply ui/src/layouts/friends.rs]: resolved in [2e1489e](https://github.com/Satellite-im/Uplink/pull/786/commits/2e1489e5bc148c365689bb6bde5cc967245460b2)

File: common/src/notifications.rs
Comment (Flemmli97) [reply common/src/notifications.rs]: problem is on linux: clicking on the notification sends the same action string `__closed` as just letting the notification fade out.  there is no distinction between them so we need an action string. so while for windows and mac it might be possible without action string the method args will still need to be like this.
the action also makes it so you need to click on a button in the notification to make it do its thing, else it just brings the app into the front (which might or might not be desirable. personally am split about this)

Comment (sdwoodbury) [reply common/src/notifications.rs]: ok, that's fine. But can  `action_id` be generated from `NotificationAction`? 
ex:
```
fn show_with_action(notification: Notification, action: NotificationAction)  {
    let action_id = format!("{}", action);
    ...
```

File: common/src/notifications.rs
Comment (sdwoodbury) [reply common/src/notifications.rs]: ok, my mistake. In that case, it may be good to just replace the function with the one you made and add an action that's a no-op in case anyone wants to send a notification that does nothing when you click it. 

 I still think we could pick a better name but that isn't so important. maybe something containing the words `desktop` and `notification`, and optionally a verb :grin:

**Comment by stavares843** [comment]
feel free to remove changes requested label when addressing the review @Flemmli97 🔨

**Review by sdwoodbury — APPROVED** [review]
unblocking the PR. 

one thing about the use of `derive_more::Display` - i believe each enumeration needs this macro: `#[display(fmt = "<some name>")]`.

**Comment by Flemmli97** [comment]
> unblocking the PR.
> 
> one thing about the use of `derive_more::Display` - i believe each enumeration needs this macro: `#[display(fmt = "<some name>")]`.

it doesnt. if that is missing it will just use the name of the enum as defined. unless we want to use some other name (as that is also used to create the translation id.

**Comment by phillsatellite** [comment]
When I tested with the most recent changes I had a couple issues, 

when you click the notification it no longer takes you to the chat 

I noticed on Windows and Linux the _open notification_ no longer appears on alert. I thought maybe thats why it doesn't take you to chat when clicked

https://github.com/Satellite-im/Uplink/assets/93608357/c78d5d47-6cbc-47dd-8701-e328e6e16fac


https://github.com/Satellite-im/Uplink/assets/93608357/7acb2507-b8c3-4308-9564-409e6e939bf9

**Comment by Flemmli97** [comment]
> > unblocking the PR.
> > one thing about the use of `derive_more::Display` - i believe each enumeration needs this macro: `#[display(fmt = "<some name>")]`.
> 
> it doesnt. if that is missing it will just use the name of the enum as defined. unless we want to use some other name (as that is also used to create the translation id.

actually was wrong about that. fixed

**Comment by Flemmli97** [comment]
> When I tested with the most recent changes I had a couple issues,
> 
> when you click the notification it no longer takes you to the chat
> 
> I noticed on Windows and Linux the _open notification_ no longer appears on alert. I thought maybe thats why it doesn't take you to chat when clicked
> 
>  macos.mov 
>  linux.mov

the problem was that the app wasnt being brought to the front.

i tried on macos and it worked for me. though this is after i changed display name thing and translation stuff so maybe that was the problem.

regarding linux (distros using gnome): 
when clicking the notification a new notification will be brought up containing "<app> is ready" which upon clicked on will bring the app to front. this is a known thing on gnome though e.g. see [https://www.reddit.com/r/gnome/comments/so4fcj/how_to_stop_all_the_application_is_ready/](https://www.reddit.com/r/gnome/comments/so4fcj/how_to_stop_all_the_application_is_ready/)
i dont think anything can be done from our end regarding that though unless anyone else has more info for that.
idk how this will behave on other distros

**Comment by phillsatellite** [comment]
When tested with the latest changes I got this error when trying to compile on Windows / MacOS on Linux it seemed to build fine and was the same behavior as before 

Screenshot 2023-06-05 at 8 37 18 AM

**Comment by Flemmli97** [comment]
> When tested with the latest changes I got this error when trying to compile on Windows / MacOS on Linux it seemed to build fine and was the same behavior as before
> 
> Screenshot 2023-06-05 at 8 37 18 AM

oops. should be fixed. as for linux does the thing i outlined in the message above happen for you?

**Comment by phillsatellite** [comment]
With the recent changes its working really well on MacOS and Windows the problem that you mentioned with linux, getting the second  *is ready*  notification is happening on Linux though. Maybe that is a something we could handle in a new ticket for the future so we dont hold this up?

**Review by phillsatellite — APPROVED** [review]

**Comment by Flemmli97** [comment]
> With the recent changes its working really well on MacOS and Windows the problem that you mentioned with linux, getting the second _is ready_ notification is happening on Linux though. Maybe that is a something we could handle in a new ticket for the future so we dont hold this up?

sounds good. though im just not sure if the linux thing is fixable on our end in the first place

**Comment by stavares843** [comment]
missing fixing conflict cc @Flemmli97 🔨

**Comment by sdwoodbury** [comment]
once conflicts are resolved, is it good to merge? LGTM

**Comment by Flemmli97** [comment]
hmm idk why but this pr is not getting updated. i merged dev already

**Comment by stavares843** [comment]
so far is updated 🎉 it has the latest dev

**Pull request merged**
